matrix-sdk/src/main/java/org/matrix/androidsdk/rest/client/MediaScanRestClient.java

```
@@ -15,47 +15,152 @@
      */
     public void scanUnencryptedFile(final String domain, final String mediaId, final ApiCallback<MediaScanResult> callback) {
-
         mApi.scanUnencrypted(domain, mediaId).enqueue(new DefaultRetrofit2CallbackWrapper<>(callback));
     }
 
     /**
      * Scan an encrypted file.
      *
      * @param encryptedMediaScanBody the encryption information required to decrypt the content before scanning it.
-     * @param callback on success callback containing a MediaScanResult object
+     * @param callback               on success callback containing a MediaScanResult object
      */
     public void scanEncryptedFile(final EncryptedMediaScanBody encryptedMediaScanBody, final ApiCallback<MediaScanResult> callback) {
+        // Encrypt encryptedMediaScanBody if the server support it
+        getServerPublicKey(new SimpleApiCallback<String>(callback) {
+            @Override
+            public void onSuccess(String info) {
+                Call<MediaScanResult> request;
+
+                // Encrypt the data, if antivirus server supports it
+                if (!TextUtils.isEmpty(info)) {
+                    try {
+                        OlmPkEncryption olmPkEncryption = new OlmPkEncryption();
+                        olmPkEncryption.setRecipientKey(info);
+
+                        String data = JsonUtils.getCanonicalizedJsonString(encryptedMediaScanBody);
+
+                        OlmPkMessage message = olmPkEncryption.encrypt(data);
+
+                        EncryptedMediaScanEncryptedBody encryptedMediaScanEncryptedBody = new EncryptedMediaScanEncryptedBody();
+                        encryptedMediaScanEncryptedBody.encryptedBodyFileInfo = new EncryptedBodyFileInfo(message);
+
+                        request = mApi.scanEncrypted(encryptedMediaScanEncryptedBody);
+                    } catch (OlmException e) {
+                        // Fallback to unencrypted body (should not happen)
```

If we failed here, we must not send the unencrypted body for security reason
Please return an error

---

ok